Update Feature-Policy to Permissions-Policy (#22439)
mdn/content: files/en-us/web/http/permissions_policy/index.md

> ---

Reviewer: @chrisdavidmills These two are quite similar, as you would expect.

-  [files/en-us/web/http/permissions_policy/index.md](https://github.com/mdn/content/pull/22439/files#diff-61ae8b1f407d7b75f635f5b128690e02028f808e83675ee427a51f9c2fcaf4c3) and 
- [files/en-us/web/http/headers/permissions-policy/index.md](https://github.com/mdn/content/pull/22439/files#diff-dd47807f3025f9e1815dd835b39f80f5ea7ef54e666d56912cbf4b998bd69242) - this doc

Just noting, I have not checked that the correct header is being linked to! I presume generally THIS is the top level doc we should normally link to?

Author: Yes, generally the former should be linked to rather than the latter, for general intro and explanation about what the feature is. I _think_ that's the one I've been linking to for such purposes ;-)

title: Permissions Policy
slug: Web/HTTP/Permissions_Policy
tags:
  - Permissions Policy
  - Permissions-Policy
  - HTTP
  - Introduction
  - Overview
  - Reference
  - Security
  - access
  - delegation
  - header
  - permission
browser-compat: http.headers.Permissions-Policy
---

{{HTTPSidebar}}{{SeeCompatTable}}

> **Permissions Policy** provides mechanisms for web developers to explicitly declare what functionality can and cannot be used on a web site. You opt-in to a set of "policies" that restrict what APIs the site can access or modify the browser's default behavior for certain features. This allows you to enforce best practices, even as the codebase evolves — as well as more safely compose third-party content.

Reviewer: All true, but this reads as though it might be something you do in your code - i.e. "opt-in" while actually it is a feature mainly of the server setup.

Author: I've updated that sentence to 

> You define a set of "policies" that restrict what APIs the site's code can access or modify the browser's default behavior for certain features.

Permissions Policy is similar to {{Glossary("CSP", "Content Security Policy")}} but controls features instead of security behavior.

Examples of what you can do with Permissions Policy:

- Change the default behavior of autoplay on mobile and third-party videos.
- Restrict a site from using sensitive devices like the camera, microphone, or speakers.
- Allow iframes to use the [Fullscreen API](/en-US/docs/Web/API/Fullscreen_API).
- Stop items from being scripted if they are not visible in the viewport, to improve performance.

> > **Note:** Permissions Policy used to be called Feature Policy. The name has changed, and so has the HTTP header syntax, so bear this in in mind if you have used Feature Policy in the past. Chrome supports the {{HTTPHeader("Permissions-Policy")}} header since version 88, with `Feature-Policy` still being supported as an alias. The `<iframe allow=" ... ">` syntax has stayed the same. Non-Chromium browsers currently don't support {{HTTPHeader("Permissions-Policy")}}, but they do support policy usage through the `allow` attribute.

Reviewer: Not sure about "Non-Chromium browsers currently don't support {{HTTPHeader("Permissions-Policy")}}, but they do support policy usage through the `allow` attribute."

It won't age well, and normally we keep this in the compat data.

Author: I've removed all the browser-specific stuff that won't age well, and directed readers towards the support tables for more information. It now reads like this:

> **Note:** Permissions Policy used to be called Feature Policy. The name has changed, and so has the HTTP header syntax, so bear this in in mind if you have used Feature Policy in the past, and check the browser support tables. The `<iframe allow=" ... ">` syntax has stayed the same.

## Concepts and usage

The web provides functionality and APIs that may have privacy or security risks if abused. In such cases, you may wish to strictly limit how functionality is used on a website. In each case, there should be an intuitive or non-breaking way for web developers to detect and handle cases where a feature is disabled.

Some approaches include:

- "Permission denied" is returned for JavaScript APIs that require user permission grants.
- JavaScript APIs that provide access to features return `false` or throw an error.

> - Options that control the feature behavior have different default values.

Reviewer: Perhaps also "APIs are not even exposed" - as though they did not exist.

Author: This is pretty closely related to the second bullet, but it isn't the same; I've put the following underneath it:

> APIs are not even exposed, as though they don't exist.

> **Note:** Newly-introduced features may have an explicit API to signal the state. Existing features that later integrate with Permissions Policy will typically use existing mechanisms.

> Permissions Policy allows you to control which origins can use which features, both on the top-level page and in embedded {{htmlelement("iframe")}}s. The aim is to enforce best practices for good user experiences and provide granular control over _sensitive_ or [powerful features](https://w3c.github.io/permissions/#dfn-powerful-feature) (meaning features that a user is required to give express permission for usage of, before related code can be executed).

Reviewer: Linking to spec definitions frowned upon, and probably not needed

```suggestion
Permissions Policy allows you to control which origins can use which features, both on the top-level page and in embedded {{htmlelement("iframe")}}s. The aim is to enforce best practices for good user experiences and provide granular control over _sensitive_ or _powerful_ features (meaning features that a user is required to give express permission for usage of, before related code can be executed).
```

To control each feature, you write a policy that consists of:

- A **directive** that identifies the name of the feature to control. See the [list of the different available directives](/en-US/docs/Web/HTTP/Headers/Permissions-Policy#directives).
- An **allowlist**, which is a list of origins that the feature should be controlled in. You can enable a feature for all or specific origins, or block its usage in all origins.

Permissions Policy provides two ways to specify policies:

- The {{httpheader("Permissions-Policy")}} HTTP header, to control feature usage in received responses and any embedded content within the page (which includes {{htmlelement("iframe")}}s).

> - The {{htmlelement("iframe")}} [`allow`](/en-US/docs/Web/HTML/Element/iframe#attributes) attribute, to control feature usage only in specific {{htmlelement("iframe")}}s.

Reviewer: OK, so somewhere here we need to define the inheritance model. At this point I read this as saying there is no relationship between what happens in the HTTP header and the iframe - you can do either or both. But I believe from the other docs, the iframe can only restrict access within the bounds already set for the page.

Author: I've added a link to the section that explains the relationship.

> Scripts can programmatically query information about the feature policy via the {{DOMxRef("FeaturePolicy")}} object located at either {{DOMxRef("Document.featurePolicy")}} or {{DOMxRef("HTMLIFrameElement.featurePolicy")}}.

Reviewer: ```suggestion
Scripts can programmatically query information about the permission policy via the {{DOMxRef("FeaturePolicy")}} object located at either {{DOMxRef("Document.featurePolicy")}} or {{DOMxRef("HTMLIFrameElement.featurePolicy")}}.
```

## Relationship with the Permissions API

Permissions Policy and the {{domxref("Permissions API", "Permissions API", "", "nocode")}} are closely-related, but different. The features that have their permissions controlled by both these technologies overlap.

> - Permissions Policy allows developers to specify whether a feature can be used in a particular document (or embeded `<frame>`s within it). These are referred to as **policy-controlled** features — see the [list of Permissions Policy directives](/en-US/docs/Web/HTTP/Headers/Permissions-Policy#directives).
> - The Permissions API provides a consistent, programmatic way to query the status of user-granted API permissions for powerful features attributed to the current document. For example, if permission to access a particular API has been granted or denied. These features are recorded in the [Permissions Registry](https://w3c.github.io/permissions-registry/).

Reviewer: I'm not sure, but I see the difference as server-granted vs user granted. It is true that a developer might set the permissions policy, but probably more likely the webmaster right? 

```suggestion
- Permissions Policy allows a server to set whether a feature can be used in a particular document (or embedded `<frame>`s within it). These are referred to as **policy-controlled** features — see the [list of Permissions Policy directives](/en-US/docs/Web/HTTP/Headers/Permissions-Policy#directives).
- The Permissions API gates access to features based on user-granted permissions. These features are recorded in the [Permissions Registry](https://w3c.github.io/permissions-registry/).
```

The identifying string used for each feature is kept consistent across both, for example, `geolocation` for the {{domxref("Geolocation API", "Geolocation API", "", "nocode")}}. Most of the API features in the Permissions Registry also have a corresponding Permissions Policy directive. One exception is the {{domxref("Notifications API", "Notifications API", "", "nocode")}}.

> In practical terms, when you set a Permissions Policy that blocks the use of a powerful feature, the user won't be asked for permission to use it, and the Permissions API {{domxref("Permissions.query", "query()")}} method will always return a {{domxref("PermissionStatus.state", "state")}} value of `denied`.

Reviewer: ```suggestion
Generally when a Permissions Policy blocks the use of a powerful feature, the user won't even be asked for permission to use it, and the Permissions API {{domxref("Permissions.query", "query()")}} method will return a {{domxref("PermissionStatus.state", "state")}} value of `denied`.
```

See also [Permissions > Relationship to the Permissions Policy specification](https://w3c.github.io/permissions/#relationship-to-permissions-policy).

## allowlists

> An allowlist is a list of origins that takes one or more of the following values contained in parentheses, separated by spaces:

Reviewer: I am not sure if it is "allow list", "allow list" or "allow-list". We should agree this across MDN though.

Reviewer: Also this feels a little premature- we haven't even said that you specify permissions with the permission policy header.
This might make sense, so leaving for you to think if it is really where you want it.

Author: In the spec, it is "allowlist", so I've tried to use this everywhere in my work. I'll do a `git grep` now, and make fixes as needed.

Good call on the ordering here. I've moved things around a bit so that it makes more sense.

Reviewer: Yes, this structure works for me!

- `src`: The feature will be allowed in this `<iframe>`, as long as the document loaded into it comes from the same origin as the URL in its {{HTMLElement('iframe','src','#Attributes')}} attribute. This value is only used in the `<iframe>` `allow` attribute, and is the _default_ `allowlist` value in `<iframe>`s.
- `"<origin>"`: The feature is allowed for specific origins (for example, `"https://a.example.com"`). Origins should be separated by spaces. Note that origins in `<iframe>` allow attributes are not quoted.

The values `*` and `()` may only be used on their own, while `self` and `src` may be used in combination with one or more origins.

> **Note:** Directives have a default allowlist, which is always one of `*`, `self`, or `none` for the `Permissions-Policy` HTTP header, and governs the default behavior if they are not explicitly listed in a policy. These are specified on the individual [directive reference pages](/en-US/docs/Web/HTTP/Headers/Permissions-Policy#directives). For `<iframe>` `allow` attributes , the default behavior is always `src`.

> Chrome 108 introduced the ability to include wildcards in Permissions Policy origins. This means that instead of having to explicitly specify several different subdomains in an allowlist, you can specify them all in a single origin with a wildcard.

Reviewer: Version information in BCD only.

Author: updated

allowlist examples:

- `*`
- `()`
- `(self)`
- `(src)`
- `("https://a.example.com")`
- `("https://a.example.com" "https://b.example.com")`
- `(self "https://a.example.com" "https://b.example.com")`
- `(src "https://a.example.com" "https://b.example.com")`
- `("https://*.example.com")`

## Permissions-Policy header syntax

The general syntax looks like this:

``` http
Permissions-Policy: <feature name>=<allowlist>
```

So for example to block all access to geolocation, you would do this:

```http
Permissions-Policy: geolocation=()
```

Or to allow access to a subset of origins, you'd do this:

```http
Permissions-Policy: geolocation=(self "https://a.example.com" "https://b.example.com")
```

Several features can be controlled at the same time by sending the header with a comma-separated list of policies, or by sending a separate header for each policy.

For example, the following are equivalent:

```http
Permissions-Policy: picture-in-picture=(), geolocation=(self https://example.com), camera=*;

Permissions-Policy: picture-in-picture=()
Permissions-Policy: geolocation=(self https://example.com)
Permissions-Policy: camera=*
```

> ## &lt;iframe&gt; syntax

Reviewer: There are comments in the other header docs on this. Perhaps map whatever you do there (if anything) here.

Author: I've removed the escaped braces.

Several features can be controlled at the same time by including a semi-colon-separated list of policy directives inside the `allow` attribute.

```html
<iframe src="https://example.com" allow="geolocation 'self' https://a.example.com https://b.example.com; fullscreen 'none'">
```

It is worth giving the `src` value a special mention. We mentioned above that using this allowlist value will mean that the associated feature will be allowed in this `<iframe>`, as long as the document loaded into it comes from the same origin as the URL in its {{HTMLElement('iframe','src','#Attributes')}} attribute. This value is the _default_ `allowlist` value for features listed in `allow`, so the following are equivalent:

```html
<iframe src="https://example.com" allow="geolocation 'src'">

<iframe src="https://example.com" allow="geolocation">
```

> **Note:** As you'll have noticed, the syntax for `<iframe>` policies is a bit different to the syntax for `Permissions-Policy` headers. The former still uses the same syntax as the older Feature Policy specification, which was superceded by Permissions Policy.

> ## Inheritance of policies for embedded content

Reviewer: Ah, here is our inheritance info! We need to link here, in particular from the first mentions of iframes.

Author: done
